Add new prompt as output directory feature

BlueAmulet opened the pull request:
This is based on and replaces #126

Added a new feature enabled via --prompt_as_dir or -p, separates generated images into subdirectories based on the prompt.
One concern about this however is the dream_log.txt file, currently for each prompt it makes a new one in each folder.

Changed 'directory_name_from_prompt' to 'prompt_as_dir' since it is simpler, can be changed back
Single directory and multi directory options are now always sanitized
Components are truncated to 255 characters to avoid windows pathing issues
Regex is compiled cached (Though I think python also caches regex itself anyway, cleaner at least)
Minor code cleanups

blessedcoolant commented:
There's some glaring issues with this .. especially on Windows.

When the folder name is too long, the classic long name problem returns where a folder cannot be renamed or deleted. There's a hack job way to do it through robocopy which most users won't know and will freak out. So the name of the folder has to be A LOT smaller. It should never be that big and prompts can get really big.

Quite frankly I am not a fan of these prompt names as folder names. I think a more elegant solution to this problem is to let users pass a -p to the dream prompt and pick a folder name to save to. And if they don't pass one, it can save to the default prompt directory. I think that's a far better approach to this problem.

But if people really want prompt as folder names, then they need to be 10-15 characters long. Not more.

BlueAmulet commented:
Right I honestly forgot about that issue, it can be disabled by enabling LongPathsEnabled if you're on Windows 10. I've had that enabled for years now at this point without many issues.

blessedcoolant commented:
> Right I honestly forgot about that issue, it can be disabled by enabling LongPathsEnabled if you're on Windows 10. I've had that enabled for years now at this point without many issues.

Yes but its not on by default and the majority of the people won't have it. So we should go by that.

BlueAmulet commented:
Hopefully addressed now:
out of 260 characters, we subtract '######.##########.##.png', the length of the normal output dir, 2 path separators, and a NUL, giving 233 characters.

blessedcoolant commented:
> Hopefully addressed now: out of 260 characters, we subtract '######.##########.##.png', the length of the normal output dir, 2 path separators, and a NUL, giving 233 characters.

Thanks for the update. This fixes the folder name length issue but I found another problem.

If I use characters like @@@ in the prompt, the folder is created but the images fail to save. 
`[WinError 3] The system cannot find the path specified:`

Also what does setting -p to simple actually do.. ? I'm unable to find a case where the underscores are being created.

BlueAmulet commented:
> If I use characters like @@@ in the prompt, the folder is created but the images fail to save.
Will look into this.

> Also what does setting -p to simple actually do.. ? I'm unable to find a case where the underscores are being created.
The characters being filtered out are currently `<>:"/\|?*` which was the complete list on MSDN, try using one of those.

lstein reviewed (changes requested):
I'm uncomfortable with hard-coding the filename and path length maximum values, because these vary both from operating system to operating system as well as filesystem type to filesystem type. This page https://www.tutorialspoint.com/python/os_pathconf.htm recommends the use of os.pathconf() to get the system-dependent values. You should run it on the output directory, not on the root directory as shown in the example, as the two may not share the same filesystem.

BlueAmulet commented:
os.pathconf is Unix only.

lstein commented:
> os.pathconf is Unix only.
> 
> https://docs.python.org/3/library/os.html#os.pathconf
> 
> > Availability: Unix.
> 
> That being said, I'll use it if it's available and assume 255 and 260 if not.
> 
> EDIT: Implemented the use of os.pathconf, does need to be tested by someone on Linux though.
> 
> It is a bit odd though: ext4 and ntfs3g both gave me 255 for name length as expected, but vfat gave me 1530, and msdos gave me 72?

Oh dear. I didn't realize it was going to be this convoluted. vfat is definitely limited to 255 characters for a filename, and distant unhappy memories suggest that msdos is limited to 8.3. You sure you're checking max filename and not max filepath?

Maybe the more general solution is to check for an IOError exception when creating the directory and file and back out gracefully.

BlueAmulet commented:
Did a rough fuzz of my sanitizer against pathvalidate and found that path components can't end in spaces or dots, so added that in. That was likely the reason of the `[WinError 3]` mentioned by blessedcoolant

One of the reasons I'm not just going with pathvalidate though is, it has some undesirable behavior regarding absolute paths, drive colons, and '..'. It also sometimes makes an absolute path out of one that wasn't before, which will break joining to the base directory.

>You sure you're checking max filename and not max filepath?

Yeah, booted into linux for a moment and was testing the output of PC_NAME_MAX on mount points, PC_PATH_MAX was always 4096.

lstein reviewed (approved):
Let's go for it.

bakkot commented on scripts/dream.py:
```diff
@@ -167,6 +177,23 @@ def main_loop(t2i, outdir, parser, infile):
+                subdir = os.path.join(*[path_filter.sub('_', x)[:name_max].rstrip(' .') for x in opt.prompt.replace('\\', '/').split('/') if x.rstrip(' .') != ''])
+
+            # truncate path to maximum allowed length
+            subdir = subdir[:(path_max - 27 - len(os.path.abspath(outdir)))]
```
Where is `27` coming from?

BlueAmulet replied:
Addressed in a previous pr comment, it is the length of '######.##########.##.png', plus two directory separators and a nul terminator.

bakkot commented on scripts/dream.py:
```diff
@@ -167,6 +177,23 @@ def main_loop(t2i, outdir, parser, infile):
+            if prompt_as_dir == 'simple':
+                # this line will sanitize the entire prompt as one folder
+                subdir = path_filter.sub('_', opt.prompt)[:name_max].rstrip(' .')
+            else:
```
This should error if the value isn't `simple` or `y`, lest someone try to put `-p` before the start of their prompt and get confused why it's missing the first word.

BlueAmulet replied:
I'm honestly not a fan of this system cause it could lead to exactly what you said, it was mostly copied over from the previous PR.

bakkot commented on scripts/dream.py:
```diff
@@ -167,6 +177,23 @@ def main_loop(t2i, outdir, parser, infile):
+                # this line will sanitize the entire prompt as one folder
+                subdir = path_filter.sub('_', opt.prompt)[:name_max].rstrip(' .')
+            else:
+                # this line will allow the prompt to create subdirectories
```
I don't think we should allow the prompt to create subdirectories. If I ask for `a human/dragon hybrid` or something, I would not expect to get a directory named "a human" containing a directory named "dragon hybrid".

BlueAmulet replied:
Somewhat agree, I don't personally like the strip/preserve folders choice cause it's a little clunky and could lead to someone accidentally dropping the first word of their prompt as mentioned before. Wondering what others think about this.

bakkot commented on scripts/dream.py:
```diff
@@ -320,6 +347,14 @@ def create_argv_parser():
+        type=str,
+        nargs='?',
+        const='y',
+        help='Output images are placed in subdirectories named using the prompt string. Only used if --prompt_as_dir or -p is specified.\nIf set to \"simple\" then all non-alphanumeric characters in the prompt will be replaced by underscores for the folder name.',
```
I'm not sure what "Only used if --prompt_as_dir or -p is specified" is trying to say here? All flags are only used if specified; that's how flags work.

BlueAmulet replied:
Was copied over from the previous PR, should probably be dropped yeah.

bakkot reviewed (commented):
Couple questions. Also I think I agree with @blessedcoolant that a flag which allowed you to explicitly specify the directory is probably more useful than trying to infer the name from the prompt - when you're iterating on the prompt you'll end up with a lot of subdirectories, instead of having one directory for all of your iterations (which you'd then change when you started working on something else).

BlueAmulet commented:
Addressed the concerns that bakkot raised up. Made it as a separate commit in case the old with/without subdirectories design was wanted. There is already an option --outdir -o that allows one to specify where the images should go, this is just to add in an automatic option.

bakkot commented:
LGTM but with the current design I might name this `--outdir-from-prompt` or something? Given that `--outdir` is the current way of specifying it.

(Also there should probably be an error if you specify both?)
